Sub-service / nested-leg failure can close the parent connection's MultiChannel

## Summary

A nested sub-service leg dying appears to close the **entire parent connection's** `MultiChannel`, not just the sub-channel — taking down unrelated services multiplexed on the same WebSocket. Surfaced live from konstructor's LSP bridge.

## Observed symptom (konstructor, 2026-06-10, adolin)

konstructor serves `Konstructor` over one ksrpc WebSocket (`ServiceApp`). `KonstructionService.lsp(client)` returns a `BridgeLanguageServer` sub-service **and** accepts a reverse-channel `KsrpcLanguageClient` param — both ride the same connection. There is also a separate `SingleChannelConnection` from the backend to a `kotlin-lsp` subprocess (the LSP engine).

When the kotlin-lsp **subprocess crashed** mid-session, the **frontend** connection's `MultiChannel` ended up **closed**. After that, every normal `KonstructionService` call (`getInfo`, `requestKonstructs`) and every `KonstructionListener` reverse push failed with:

```
java.lang.IllegalStateException: MultiChannel (com.monkopedia.ksrpc.internal.MultiChannel@…) is closed
Caused by: java.util.concurrent.CancellationException: Multi-channel failure
```

Net user impact: the editor silently stopped executing builds entirely — saves compiled but never ran, because the frontend's forward calls started throwing.

Per `ksrpc-core/src/commonMain/kotlin/internal/MultiChannel.kt`, `MultiChannel` is **per-connection** (request/response correlation for one transport), so "MultiChannel is closed" == the whole connection is down, not a single sub-channel.

## Question

Should closing/failing a returned sub-service (or a param-passed reverse-channel service) ever close the **parent** connection's `MultiChannel`? Specifically:

1. Is there a `SubServiceChannel` / `serviceMap` teardown path where a sub-channel `close()` (or an exception escaping a sub-service method) cascades up to close the connection-level `MultiChannel`?
2. Should an exception thrown by a sub-service method be isolated to that call's error response, rather than tearing down the transport?

If ksrpc can guarantee **sub-service close ⊄ connection close**, that's the durable fix for every consumer — konstructor is one, but any app multiplexing independent services over one WebSocket is exposed.

## Konstructor-side mitigation already shipped

konstructor added a guard (`BridgeLanguageServer.guardEngine`) wrapping every forward call to the LSP subprocess so a crashed engine degrades to inert instead of throwing onto the shared connection. That stops the bleeding for konstructor, but the framework-level isolation question stands.

Happy to share the full `BridgeLanguageServer` close/teardown + adolin logs. Filed by the konstructor repo agent on behalf of a live debugging session; not release-blocking (workaround exists), but it's a real multiplexing fragility.

--- a/ksrpc-core/src/commonMain/kotlin/internal/HostSerializedServiceImpl.kt
+++ b/ksrpc-core/src/commonMain/kotlin/internal/HostSerializedServiceImpl.kt
@@ -34,9 +34,11 @@ import com.monkopedia.ksrpc.channels.RpcCallId
 import com.monkopedia.ksrpc.channels.SerializedChannel
 import com.monkopedia.ksrpc.channels.SerializedService
 import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.coroutineContext
 import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.CompletableDeferred
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.ensureActive
 import kotlinx.coroutines.withContext
 import kotlinx.serialization.builtins.serializer
 
@@ -84,7 +86,20 @@ class HostSerializedChannelImpl<T>(
             }
         }
     } catch (t: CancellationException) {
-        throw t
+        // A CancellationException here is genuine cooperative cancellation ONLY if THIS
+        // coroutine's job was actually cancelled (the caller went away, or an incoming cancel
+        // frame cancelled the handler) — in that case ensureActive() rethrows it and we
+        // propagate as before. Otherwise it is a *foreign* cancellation that leaked out of a
+        // downstream call — e.g. a different connection's MultiChannel.close completing its
+        // pending awaits with CancellationException, surfaced here through a bridging
+        // sub-service. A foreign cancellation must NOT propagate into this connection's
+        // receive loop, where it would close this connection's MultiChannel and take down
+        // every other service multiplexed on it. Treat it like any other dispatch failure:
+        // error-frame this one call so it fails while the connection survives (#228).
+        coroutineContext.ensureActive()
+        env.logger.info("SerializedChannel", "Foreign cancellation during dispatching", t)
+        env.errorListener.onError(t)
+        CallData.Error(KsrpcException.INTERNAL_ERROR_CODE, t.message ?: t.toString())
     } catch (t: Throwable) {
         env.logger.info("SerializedChannel", "Exception thrown during dispatching", t)
         env.errorListener.onError(t)

--- a/ksrpc-test/src/commonTest/kotlin/RpcForeignCancellationTest.kt
+++ b/ksrpc-test/src/commonTest/kotlin/RpcForeignCancellationTest.kt
@@ -1,0 +1,104 @@
+/*
+ * Copyright (C) 2026 Jason Monk <monkopedia@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.monkopedia.ksrpc
+
+import com.monkopedia.ksrpc.annotation.KsMethod
+import com.monkopedia.ksrpc.annotation.KsService
+import com.monkopedia.ksrpc.channels.registerDefault
+import com.monkopedia.ksrpc.sockets.asConnection
+import io.ktor.utils.io.close
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFails
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withTimeout
+
+@KsService
+interface ForeignCancelInterface : RpcService {
+    @KsMethod("/normal")
+    suspend fun normal(input: String): String
+
+    @KsMethod("/foreign_cancel")
+    suspend fun foreignCancel(input: String): String
+}
+
+/**
+ * Regression test for #228: a foreign [CancellationException] thrown by a hosted service method
+ * — modelling a downstream/peer connection dying and surfacing its `MultiChannel.close`
+ * cancellation through a bridging sub-service — must NOT tear down the hosting connection. It
+ * should be isolated to that one call's error response, leaving every other service multiplexed
+ * on the same connection still callable.
+ *
+ * Before the fix, `HostSerializedChannelImpl.call` rethrew every [CancellationException]
+ * (including foreign ones), which propagated into the hosting connection's receive loop and
+ * closed its `MultiChannel` — so the second, unrelated call below failed with
+ * "MultiChannel is closed".
+ */
+class RpcForeignCancellationTest {
+
+    @Test
+    fun foreignCancellationDoesNotTearDownConnection() = runBlockingUnit {
+        if (RpcFunctionalityTest.TestType.PIPE !in RpcFunctionalityTest.TestType.values()) {
+            return@runBlockingUnit
+        }
+        val (output, input) = createPipe()
+        val (so, si) = createPipe()
+        val serviceChannel = (si to output).asConnection(
+            ksrpcEnvironment {
+                errorListener = ErrorListener { /* expected: the foreign cancellation */ }
+            }
+        )
+        val clientChannel = (input to so).asConnection(
+            ksrpcEnvironment {
+                errorListener = ErrorListener { /* swallow for the test */ }
+            }
+        )
+        val bgJob = GlobalScope.launch(Dispatchers.Default) {
+            serviceChannel.registerDefault<ForeignCancelInterface, String>(
+                object : ForeignCancelInterface {
+                    override suspend fun normal(input: String): String = "ok: $input"
+
+                    override suspend fun foreignCancel(input: String): String =
+                        throw CancellationException("Multi-channel failure")
+                }
+            )
+        }
+        try {
+            withTimeout(30_000) {
+                val service = clientChannel.defaultChannel().toStub<ForeignCancelInterface, String>()
+                // 1. The foreign-cancellation call surfaces as an error to the caller...
+                assertFails { service.foreignCancel("boom") }
+                // 2. ...but the connection survives: an unrelated call still works.
+                assertEquals("ok: hello", service.normal("hello"))
+            }
+        } finally {
+            try {
+                input.cancel(null)
+            } catch (t: Throwable) {
+            }
+            try {
+                si.cancel(null)
+            } catch (t: Throwable) {
+            }
+            output.close(null)
+            so.close(null)
+            bgJob.join()
+        }
+    }
+}

--- a/ksrpc-test/src/commonTest/kotlin/RpcForeignCancellationTest.kt
+++ b/ksrpc-test/src/commonTest/kotlin/RpcForeignCancellationTest.kt
@@ -81,7 +81,8 @@ class RpcForeignCancellationTest {
         }
         try {
             withTimeout(30_000) {
-                val service = clientChannel.defaultChannel().toStub<ForeignCancelInterface, String>()
+                val service =
+                    clientChannel.defaultChannel().toStub<ForeignCancelInterface, String>()
                 // 1. The foreign-cancellation call surfaces as an error to the caller...
                 assertFails { service.foreignCancel("boom") }
                 // 2. ...but the connection survives: an unrelated call still works.

